fluent-langneg/src/index.js

```diff
@@ -0,0 +1,266 @@
+ *
+ *    What it does ([requested] * [available] = [supported]):
+ *
+ *    ['fr'] * ['fr-FR'] = ['fr-FR']
+ *    ['en'] * ['en-US'] = ['en-US']
+ *    ['sr'] * ['sr-Latn', 'sr-Cyrl'] = ['sr-Cyrl']
+ *
+ *  3) region range check has been added
+ *
+ *    Lastly, the last form of check is against the requested locale ID
+ *    but with the region field replaced with a `*` range.
+ *
+ *    The rationale here laid out in LDML 4.4 Language Matching:
+ *      "(...) normally the fall-off between the user's languages is
+ *      substantially greated than regional variants."
+ *
+ *    In other words, if we can't match for the given region, maybe
+ *    we can match for the same language/script but other region, and
+ *    it will in most cases be preferred over falling back on the next
+ *    language.
+ *
+ *    What it does ([requested] * [available] = [supported]):
+ *
+ *    ['en-AU'] * ['en-US'] = ['en-US']
+ *    ['sr-RU'] * ['sr-Latn-RO'] = ['sr-Latn-RO'] // sr-RU -> sr-Latn-RU
+ */
+function filterMatches(requestedLocales, availableLocales, likelySubtags) {
+  const supportedLocales = [];
+
+  outer:
+  for (const requestedLocale of requestedLocales) {
+
+    for (const availableLocale of availableLocales) {
+      if (requestedLocale === availableLocale) {
+        supportedLocales.push(availableLocale);
+        continue outer;
+      }
+    }
+
+    for (const availableLocale of availableLocales) {
+      if (localesEqual(requestedLocale, availableLocale)) {
+        supportedLocales.push(availableLocale);
+        continue outer;
+      }
+    }
+
+    const maxRequestedLocale = addLikelySubtags(requestedLocale, likelySubtags);
+
+    if (maxRequestedLocale) {
+      for (const availableLocale of availableLocales) {
+        if (localesEqual(maxRequestedLocale, availableLocale)) {
+          supportedLocales.push(availableLocale);
+          continue outer;
+        }
+      }
+    }
+
+    const regionRange = regionRangeFor(maxRequestedLocale || requestedLocale);
+
+    for (const availableLocale of availableLocales) {
+      if (localesEqual(regionRange, availableLocale)) {
+        supportedLocales.push(availableLocale);
+        continue outer;
+      }
+    }
+  }
+
+  return supportedLocales;
+}
+
+function GetOption(options, property, type, values, fallback) {
+  let value = options[property];
+
+  if (value !== undefined) {
+    if (type === 'boolean') {
+      value = new Boolean(value);
+    } else if (type === 'string') {
+      value = value.toString();
+    }
+
+    if (values !== undefined && values.indexOf(value) === -1) {
+      throw new Error('Invalid option value');
+    }
+
+    return value;
+  }
+
+  return fallback;
+}
+
+export default function negotiateLanguages(requestedLocales,
+                                   availableLocales,
+                                   options = {}) {
+
+  const defaultLocale = GetOption(options, 'defaultLocale', 'string');
```

Comment:
I don't think it is. Since the algorithm works perfectly fine without it, I'd prefer to keep only the necessary arguments as positional - requestedLocales and availableLocales.